pkg/readiness/ready_tracker.go

```diff
@@ -192,11 +193,126 @@ func (t *Tracker) Run(ctx context.Context) error {
 	return nil
 }
 
+// collectForObjectTracker identifies objects that are unsatisfied for the provided
+// `es`, which must be an objectTracker, and removes those expectations
+func (t *Tracker) collectForObjectTracker(ctx context.Context, es Expectations) error {
+	if es == nil {
+		return fmt.Errorf("nil Expectations provided to collectForObjectTracker")
+	} else if !es.Populated() || es.Satisfied() {
+		log.Info("Expectations unpopulated or already satisfied, skipping collection")
+		return nil
+	}
+
+	// es must be an objectTracker so we can fetch `unsatisfied` expectations and get GVK
+	ot, ok := es.(*objectTracker)
+	if !ok {
+		return fmt.Errorf("Expectations was not an objectTracker in collectForObjectTracker")
+	}
+
+	// there is only ever one GVK for the unsatisfied expectations of a particular objectTracker.
+	gvk := ot.gvk
+	ul := &unstructured.UnstructuredList{}
+	ul.SetGroupVersionKind(gvk)
+	if err := t.lister.List(ctx, ul); err != nil {
+		return errors.Wrapf(err, "while listing %v in collectForObjectTracker", gvk)
+	}
+
+	// identify objects in `unsatisfied` that are were not found above.
+	// The expectations for these objects should be collected since the
+	// tracker is waiting for them, but they no longer exist.
+	unsatisfied := ot.unsatisfied()
+	unsatisfiedmap := make(map[objKey]struct{})
+	for _, o := range unsatisfied {
+		unsatisfiedmap[o] = struct{}{}
+	}
+	for _, o := range ul.Items {
+		k, err := objKeyFromObject(&o)
+		if err != nil {
+			return errors.Wrapf(err, "while getting key for %v in collectForObjectTracker", o)
+		}
+		// delete is a no-op if the key isn't found
+		delete(unsatisfiedmap, k)
+	}
+
+	// now remove the expectations for deleted objects
+	for k := range unsatisfiedmap {
+		u := &unstructured.Unstructured{}
+		u.SetName(k.namespacedName.Name)
+		u.SetNamespace(k.namespacedName.Namespace)
+		u.SetGroupVersionKind(k.gvk)
+		ot.CancelExpect(u)
+	}
+
+	return nil
+}
+
+// collectInvalidExpectations searches for any unsatisfied expectations
+// for this tracker for which the expected object has been deleted, and
+// cancels those expectations.
+// Errors are handled and logged, but do not block collection for other trackers.
+func (t *Tracker) collectInvalidExpectations(ctx context.Context) {
+	tt := t.templates
+	err := t.collectForObjectTracker(ctx, tt)
+	if err != nil {
+		log.Error(err, "while collecting for the ConstraintTemplate tracker")
+	}
+
+	ct := t.config
+	err = t.collectForObjectTracker(ctx, ct)
+	if err != nil {
+		log.Error(err, "while collecting for the Config tracker")
+	}
+
+	// collect deleted but expected constraints
+	for _, gvk := range t.constraints.Keys() {
+		// retrieve the expectations for this key
+		es := t.constraints.Get(gvk)
+		err = t.collectForObjectTracker(ctx, es)
+		if err != nil {
+			log.Error(err, "while collecting for the Constraint type", "gvk", gvk)
+			continue
```

Comment:
Should we rate limit / backoff when we see errors? If so to keep the code simpler, we might consider baking it into the lister.

Reply:
I added a wrapper for the lister that does exponential backoff...and then realized that essentially duplicates the retryLister that was already used elsewhere in this file. Do you think we should just add the retryLister to the underlying List call, or do you have something else in mind?

Reply:
I changed this to just using the retryLister. Let me know if you had something else in mind.

I think a certain amount of low-investment retrying makes sense here to cover common transient issues -- "low-investment" as this task is doing non-critical work (cleaning up to avoid a race condition).